hash.c

```diff
@@ -1873,6 +1873,51 @@ rb_hash_s_try_convert(VALUE dummy, VALUE hash)
     return rb_check_hash_type(hash);
 }
 
+/*
+ *  call-seq:
+ *     Hash.ruby2_keywords?(hash) -> true or false
+ *
+ *  Check if a given hash is flagged by Module#ruby2_keywords (or
+ *  Proc#ruby2_keywords) or not.
+ *  This method is not for casual use; debugging, researching, and
+ *  some truly necessary cases like serialization of arguments.
+ *
+ *     ruby2_keywords def foo(*args)
+ *       Hash.ruby2_keywords?(args.last)
+ *     end
+ *     foo(k: 1)   #=> true
+ *     foo({k: 1}) #=> false
+ */
+static VALUE
+rb_hash_s_ruby2_keywords_p(VALUE dummy, VALUE hash)
+{
+    Check_Type(hash, T_HASH);
+    return (RHASH(hash)->basic.flags & RHASH_PASS_AS_KEYWORDS) ? Qtrue : Qfalse;
+}
+
+/*
+ *  call-seq:
+ *     Hash.ruby2_keywords!(hash) -> hash
+ *
+ *  Add a ruby2_keywords flag to a given hash.
+ *  This method is not for casual use; debugging, researching, and
+ *  some truly necessary cases like deserialization of arguments.
+ *
+ *     h = {k: 1}
+ *     Hash.ruby2_keywords!(h)
+ *     def foo(k: 42)
+ *       k
+ *     end
+ *     foo(*[h]) #=> 1 with neither a warning nor an error
+ */
+static VALUE
+rb_hash_s_ruby2_keywords_bang(VALUE dummy, VALUE hash)
+{
+    Check_Type(hash, T_HASH);
+    RHASH(hash)->basic.flags |= RHASH_PASS_AS_KEYWORDS;
```

Comment:
Should this check if the Hash is frozen?

Reply:
Ah, good point.  Should we raise an exception?

Reply:
I think so, yes. But I would prefer if it was not mutating at all.